Missing buttons and errors when deleting or editing a comment

StanleyGoldman opened the pull request:
- Correctly utilizing an ObservableAsPropertyHelper
  - As part of a last minute fix to #1701 I attempted to use `ToProperty` but I completely failed on the follow through.
- Standalone comments made by the rest api do not return a node Id, the lack of this data results in the deletion or editing of standalone comments fail in error, in some cases.
  - We need to use Graphql to make standalone comments. In order to fix that some code from #1712 was ported over

meaghanlewis commented:
I'm not able to update comments
![screen shot 2018-06-08 at 2 56 58 pm](https://user-images.githubusercontent.com/10103121/41182874-7f65df1e-6b2c-11e8-85c8-738e6ffb8219.png)

and see this error in the log:
```2018-06-08 14:55:09.434 [04640] INFO [20] GitHubPackage             Initializing GitHub Extension v2.5.3.2815 in Microsoft Visual Studio 2015 (14.0.25420.1)
2018-06-08 14:56:50.896 [04640] EROR [01] CommentViewModel          Error posting comment
System.AggregateException: One or more errors occurred. ---> Octokit.GraphQL.Core.GraphQLQueryException: Argument 'input' on Field 'updatePullRequestReviewComment' has an invalid value. Expected type 'UpdatePullRequestReviewCommentInput!'.
   --- End of inner exception stack trace ---
   at Octokit.GraphQL.Core.Deserializers.ResponseDeserializer.Deserialize[TResult](CompiledQuery`1 query, JObject data)
   at Octokit.GraphQL.Core.Deserializers.ResponseDeserializer.Deserialize[TResult](CompiledQuery`1 query, String data)
   at Octokit.GraphQL.Connection.<Run>d__14`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at GitHub.InlineReviews.Services.PullRequestSessionService.<EditComment>d__31.MoveNext() in C:\projects\visualstudio\src\GitHub.InlineReviews\Services\PullRequestSessionService.cs:line 656
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<EditComment>d__21.MoveNext() in C:\projects\visualstudio\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 183
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.ViewModels.InlineCommentThreadViewModel.<DoEditComment>d__6.MoveNext() in C:\projects\visualstudio\src\GitHub.InlineReviews\ViewModels\InlineCommentThreadViewModel.cs:line 85
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.ViewModels.CommentViewModel.<DoCommitEdit>d__15.MoveNext() in C:\projects\visualstudio\src\GitHub.InlineReviews\ViewModels\CommentViewModel.cs:line 171
---> (Inner Exception #0) Octokit.GraphQL.Core.GraphQLQueryException: Argument 'input' on Field 'updatePullRequestReviewComment' has an invalid value. Expected type 'UpdatePullRequestReviewCommentInput!'.<---

---> (Inner Exception #1) Octokit.GraphQL.Core.GraphQLQueryException: Argument 'pullRequestReviewCommentId' on InputObject 'UpdatePullRequestReviewCommentInput' has an invalid value. Expected type 'ID!'.<---
```

StanleyGoldman commented:
@grokys Would you expect this to be necessary...

https://github.com/github/VisualStudio/pull/1723/files#diff-27e704a8cecf704234bf4fe243bcfc93R632

shana commented on src/GitHub.InlineReviews/Services/PullRequestSessionService.cs:
```diff
@@ -592,6 +594,8 @@ public ITextDocument GetDocument(ITextBuffer buffer)
                 DiffHunk = result.DiffHunk,
                 Id = result.Id,
+                //TODO: Populate this
+                NodeId = string.Empty,
```
Could you elaborate on why this needs to be created and why the TODO comment (I'm totally clueless on this part of the code)

grokys replied:
I think this is where Stanley got stuck. The problem is that we're currently mixing GraphQL and REST queries and each of them has a different concept of ID, so we currently need both. #1712 will fix this properly by moving to GQL for all this stuff, but in the interim I think we can convert the two "Add Comment" methods that currently use REST, to using GQL; though will be a bit messy because there's currently no way to actually do this in a single GQL mutation.

StanleyGoldman replied:
Apologies for the lack of communication here. I left my status with this in Slack.

@grokys is correct. I migrated some of his code from #1712 in order to fill the gaps.

StanleyGoldman commented on src/GitHub.InlineReviews/Services/PullRequestSessionService.cs:
```diff
@@ -528,77 +528,71 @@ public ITextDocument GetDocument(ITextBuffer buffer)
+            var graphql = await graphqlFactory.CreateConnection(address);
 
-            return new PullRequestReviewCommentModel
+            var addReview = new AddPullRequestReviewInput
```
Test comment

meaghanlewis commented:
@StanleyGoldman  I can edit/delete new comments I've made right away, but I noticed a few other things:

- I am unable to edit existing comments - no edit/delete icons shows
![cant edit existing](https://user-images.githubusercontent.com/10103121/41258289-8508d35e-6d84-11e8-930c-723a72675e34.png)

- New single comments are added as `Pending`
![new comment](https://user-images.githubusercontent.com/10103121/41258332-97fb7ee4-6d84-11e8-8edc-05833b7e69c0.gif)

- And I think because they are pending, when I try to to reply to single comments, that fails:

![reply to single comment](https://user-images.githubusercontent.com/10103121/41258344-a42dfbd8-6d84-11e8-92c7-440c0abb546c.gif)

grokys commented:
As regards @meaghanlewis issues:

> I am unable to edit existing comments - no edit/delete icons shows

I'm not seeing this. Not sure what is needed to repro.

> New single comments are added as Pending

This is quite simple: `IsPending` is [erroneously being set to `true` here](https://github.com/github/VisualStudio/blob/b58ae5b13dd4bef694cb314304f87f1677c234f4/src/GitHub.InlineReviews/Services/PullRequestSessionService.cs#L577) and needs to be set to `false` [after this call](https://github.com/github/VisualStudio/blob/b58ae5b13dd4bef694cb314304f87f1677c234f4/src/GitHub.InlineReviews/Services/PullRequestSessionService.cs#L594)

> when I try to to reply to single comments, that fails:

This is because database IDs are being used as Node IDs [here](https://github.com/github/VisualStudio/pull/1723/files#diff-27e704a8cecf704234bf4fe243bcfc93R594)

StanleyGoldman commented on src/GitHub.InlineReviews/Services/IPullRequestSessionService.cs:
```diff
@@ -286,17 +283,14 @@ public interface IPullRequestSessionService
-        /// <param name="remoteRepositoryOwner">The owner of the repository fork to post to.</param>
         /// <param name="user">The user posting the comment.</param>
-        /// <param name="number">The pull request number.</param>
+        /// <param name="pullRequestNodeId">The pull request node id.</param>
```
asdf

StanleyGoldman commented on src/GitHub.InlineReviews/Services/IPullRequestSessionService.cs:
```diff
@@ -260,9 +260,8 @@ public interface IPullRequestSessionService
-        /// <param name="remoteRepositoryOwner">The owner of the repository fork to post to.</param>
         /// <param name="user">The user posting the comment.</param>
-        /// <param name="number">The pull request number.</param>
+        /// <param name="pullRequestNodeId">The pull request node id.</param>
```
asdf

StanleyGoldman replied:
asdf

StanleyGoldman commented:
Thanks @grokys for finding the errors. Yea, the `ToString()` was a straggler from after the first time I learned the lesson.

StanleyGoldman commented:
@meaghanlewis I also am unable to repro the unable to missing delete button issue, I can pair with you on it today.

meaghanlewis commented:
@StanleyGoldman this is looking good, but 2 more things. 

1. Replies made to comments are listed as `Pending` temporarily.
![reply to single comment](https://user-images.githubusercontent.com/10103121/41309224-073c590a-6e33-11e8-98bf-47d60f89c137.gif)

2. I'm unable to delete `Pending` comments.
![cant delete pending comment](https://user-images.githubusercontent.com/10103121/41309507-cd9bc7a2-6e33-11e8-8bd0-b06cbf8a0971.gif)

I'm not sure if this is getting to be out of scope for this particular fix, so I'm okay with opening separate issues for these as well. What do you think?

StanleyGoldman commented on src/GitHub.InlineReviews/Services/PullRequestSession.cs:
```diff
@@ -138,11 +138,11 @@ public string GetRelativePath(string path)
-                    RepositoryOwner,
                     User,
-                    PullRequest.Number,
+                    pullRequestNodeId,
```
asdf

StanleyGoldman replied:
asdf

StanleyGoldman commented:
@meaghanlewis I replicated and fixed the first issue, replying to standalone comments should not show as pending anymore.

As far as your second issue goes, I added functionality to display an error directly in that view. But I do not know how to fix the error you found. I'm going to try again tomorrow morning with a fresh brain.

![image](https://user-images.githubusercontent.com/417571/41324849-1165e106-6e85-11e8-959f-217ca88f4635.png)

Also it seems you were able to identify another bug with @jcansdale in Slack last night. Would you happen to know how to reproduce the error.

grokys commented:
I suspect the inability to delete pending comments is a problem with the REST API, in that it refuses to accept the existence of pending comments. This is a problem because there's no way in GraphQL to delete comments.

We spoke to someone on the API ecosystem team last week who suggested that the REST API not working with pending comments was in error, so my feeling is that we should work with them to get this fixed.

grokys commented on src/GitHub.InlineReviews/Services/PullRequestSessionService.cs:
```diff
@@ -517,7 +517,7 @@ public ITextDocument GetDocument(ITextBuffer buffer)
                     PullRequestReviewId = x.Comment.PullRequestReview.DatabaseId.Value,
                     User = user,
-                    IsPending = true,
+                    IsPending = false,
```
Are you sure this is right? This method is `PostPendingReviewCommentReply`, so doesn't this need to be set to pending?

StanleyGoldman replied:
Thanks

sguthals commented on src/GitHub.Exports.Reactive/Services/IPullRequestSession.cs:
```diff
@@ -109,12 +109,10 @@ public interface IPullRequestSession
         /// </summary>
         /// <param name="body">The comment body.</param>
-        /// <param name="inReplyTo">The REST ID of the comment to reply to.</param>
         /// <param name="inReplyToNodeId">The GraphQL ID of the comment to reply to.</param>
```
test
